04_Scheduling/02_Scheduler.md

```diff
@@ -154,14 +154,64 @@ void schedule() {
+In the previous section we have seen how to iterate throught tasks, but that basically was just iterating through items of a list, not very useful. 
+
+#### Saving Context
+
+Once the next task to be executed has been selected, the first thing that the scheduler has to do is stop the current task, and save its current status. If you have implemented the interrupt handler following this guide you should already have the context being passed to, what we want is to pass down the context to the schedule function, but first  we need to change its signature of our schedule function, adding a new parameter: 
 
 ```c
 void schedule(cpu_status_t* context);
 ````
 
+And now add the context parameter where the function is called. Now with this new parameter we have a snapshot of the cpu just before the interrupt was fired, and what was it was doing just before it? Running our task, so what we just need to do is update the context variable of the current tas just before picking the next: 
+
+```c
+void schedule(cpu_status_t* context) {
+    tasks_list[current_executing_task_idx].context = context
+    current_executing_task_idx = (current_executing_task_idx + 1) % MAX_TASKS;
+    while(tasks_list[current_executing_task_idx] == NULL) {        
+        current_executing_task_idx = (current_executing_task_idx + 1) % MAX_TASKS;
+    }
+}
+
+```
+
+that's it, this is how how we save the status of the current executing task. 
+
+#### Restoring context
+
+After the previous task status has been saved and the next task has been selected we are now ready to make the switch. How to do it is pretty straightforward, but let's have a quick recap on how we are handling interrupts: 
+
+1. The first thing is that whenever an interrupt is fired an asm code snippet is called saves the current cpu registers status, on the stack, 
+2. Then once the status is saved, it place the current stack pointer (`rsp`) on the `rdi` register (it is the first c function parameter for x86_64 architecture) and call the interrupt handler. This parameter is our `context` variable (that is passed also to the scheduler)
+3. Then our interrupt handler function (unless we are writing an os entirely in asm, this usually is on our language of choice, in our case C)  does whatever it needs to serve the interrupt  and  the context variable. Returning a variable, according to the ABI calling convention is corresponding to placing the returning value to the `rax` register
+4. Now we are back in the asm code where `rax` value is placed on the `rsp` register, and then the context restored...
```

Comment:
why `rax`? It's obvious to us that it's the sysv x86_64 calling convention, but might be worth a quick mention in brackets.